src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java

```diff
@@ -3574,27 +3596,100 @@ void writeTime(java.sql.Timestamp value, int scale) throws SQLServerException {
     void writeDateTimeOffset(Object value, int scale, SSType destSSType) throws SQLServerException {
         GregorianCalendar calendar;
         TimeZone timeZone; // Time zone to associate with the value in the Gregorian calendar
-        long utcMillis; // Value to which the calendar is to be set (in milliseconds 1/1/1970 00:00:00 GMT)
         int subSecondNanos;
         int minutesOffset;
 
-        microsoft.sql.DateTimeOffset dtoValue = (microsoft.sql.DateTimeOffset) value;
-        utcMillis = dtoValue.getTimestamp().getTime();
-        subSecondNanos = dtoValue.getTimestamp().getNanos();
-        minutesOffset = dtoValue.getMinutesOffset();
+        /*
+         * Out of all the supported temporal datatypes, DateTimeOffset is the only datatype that doesn't
+         * allow direct casting from java.sql.timestamp (which was created from a String).
+         * DateTimeOffset was never required to be constructed from a String, but with the
+         * introduction of extended bulk copy support for Azure DW, we now need to support this scenario.
+         * Parse the DTO as string if it's coming from a CSV.
+         */
+        if (value instanceof String) {
```

Comment:
Need to check for `connection.getSendTemporalDataTypesAsStringForBulkCopy()` here too.

Reply:
The check is done before this method is called. Checking for String guarantees that that connection property is set to false, and also that the data is coming from a CSV. Near the beginning of writeColumnToTdsWriter method in SQLServerBulkCopy, this line:

`        } else if (null != serverBulkData && connection.getSendTemporalDataTypesAsStringForBulkCopy()) {
`

ensures that if getSendTemporalDataTypesAsStringForBulkCopy is true then bulkJdbcType will be varchar, which means that if getSendTemporalDataTypesAsStringForBulkCopy is false, we couldn't have come to this line.